builder.go

```diff
@@ -0,0 +1,276 @@
+package engine
+
+import (
+	"fmt"
+
+	"go.arcalot.io/lang"
+	"go.flow.arcalot.io/engine/internal/dg"
+	"go.flow.arcalot.io/engine/internal/expressions"
+	"go.flow.arcalot.io/engine/internal/yaml"
+	"go.flow.arcalot.io/engine/workflow"
+	"go.flow.arcalot.io/pluginsdk/schema"
+)
+
+// treeItem is an item in the workflow dependency tree.
+type treeItem struct {
+	Type   string
+	Item   string
+	Output string
+}
+
+const yamlTypeExpr = "!expr"
+const stepsKey = "steps"
+const outputsKey = "outputs"
+const inputKey = "input"
+
+// buildDependencyTree builds a dependency tree from the input schema and steps. The function expects the input and
+// deploy parameters to be prepared as yaml.Node instances. The returned steps contain the resolved raw data and the
+// AST nodes if an expression was found.
+//nolint:funlen
+func buildDependencyTree(inputSchema *schema.ScopeSchema, steps map[string]*workflow.PluginStep, stepSchemas map[string]schema.Step, output yaml.Node) (dg.DirectedGraph[treeItem], map[string]*workflow.PluginStep, any, error) {
```

Comment:
This func has quite four (and quite complex) return values. It would be good to bundle the first three values together in a custom struct, e.g. `type DependencyTree struct {`. What do you think?